### ISSUE TITLE
[explicit-module-boundary-types] allowNames doesn't work on function and method expressions

### ISSUE THREAD

**Issue by canac** [typescript-eslint/typescript-eslint#3033]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**
`eslintrc.js`:
```JSON
{
  "rules": {
    "@typescript-eslint/explicit-module-boundary-types": ["error", { "allowedNames": ["foo1", "foo2"] }],
  }
}
```

`main.ts`:
```ts
export function foo1() {
  return 0;
}

export const object = {
  foo2() {
    return 0;
  },
};
```

`tsconfig.json`:
```json
{
  "compilerOptions": {
    "strict": true
  }
}
```

**Expected Result**

I expect that `allowedNames` will whitelist those methods.

**Actual Result**

This rule generates an error for `foo1` and `foo2`.

    1:8  warning  Missing return type on function  @typescript-eslint/explicit-module-boundary-types
    6:3  warning  Missing return type on function  @typescript-eslint/explicit-module-boundary-types

**Additional Info**

It works correctly when using arrow functions. This code gives no errors.
```ts
export const foo1 = () => 0;
export const foo2 = () => 0;
export const object = {
  foo2,
};
```

This still errors, however.
```ts
export const object = {
  foo2: () => 0,
};
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.15.0` |
| `@typescript-eslint/parser`        | `4.15.0` |
| `TypeScript`                       | `4.1.5` |
| `ESLint`                           | `7.19.0` |
| `node`                             | `15.8.0` |
